### ISSUE TITLE
assertEqual errors if associative arrays have different domains

### ISSUE THREAD

**Issue by lucaferranti** [chapel-lang/chapel#20848]
when comparing two associative arrays with different domains, `test.assertEqual` errors instead of returning false

### Summary of Problem

when comparing two associative arrays with different domains, `test.assertEqual` errors instead of returning false

### Steps to Reproduce

**Source Code:**

```chapel
use UnitTest;

var a = ["A" => 0, "B" => 2, "C" => 3],
    b = ["B" => 2, "A" => 0, "D" => 3]; // note the different key D instead of C

proc test1(test: borrowed Test) throws {
  test.assertEqual(a, b);
}

UnitTest.main();                                                                                      ~                        
```

**Compile command:**

`chpl foo.chpl`

**Execution command:**

`./foo`

**Associated Future Test(s):**

### Configuration Information

- Output of `chpl --version`:
```
chpl version 1.28.0
  built with LLVM version 12.0.0
Copyright 2020-2022 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)
```
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_RE2: none
CHPL_LLVM: system
CHPL_AUX_FILESYS: none
```

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
```
gcc (Ubuntu 10.3.0-1ubuntu1) 10.3.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by bradcray** [comment]
@lucaferranti — Thanks for reporting this.  From what I can tell with a quick look at the code for array equals, it looks like the package's author only focused on dense rectangular arrays in their implementation of `assertArrayEqual()` here https://github.com/chapel-lang/chapel/blob/main/modules/packages/UnitTest.chpl#L427-L454.  I suspect the fix for this would be fairly straightforward.  Could you let us know whether this is a blocking issue for you or whether you just happened upon it?  Thanks!

**Comment by bradcray** [comment]
Also, recording the error I see for posterity (this is with a developer setup):

```
$CHPL_HOME/modules/internal/ChapelArray.chpl:1424: error: halt reached - zippered associative array does not match the iterated domain
Stacktrace

halt() at $CHPL_HOME/modules/standard/Errors.chpl:752
halt() at $CHPL_HOME/modules/standard/Errors.chpl:738
advance() at $CHPL_HOME/modules/internal/ChapelArray.chpl:1418
coforall_fn() at $CHPL_HOME/modules/packages/UnitTest.chpl:447
wrapcoforall_fn() at $CHPL_HOME/modules/packages/UnitTest.chpl:447
```

**Comment by lucaferranti** [comment]
It's not blocking. I encountered that while working on the exercism track for chapel (I'll push the branch and post the link in a second).

I can try to have a look at the source code and try to come up with a fix, I also think it shouldn't be too hard to fix.

My first guess would be to add to [these checks] in `assertEqualArray`

```chapel
else if array1.domain != array2.domain {
  const errorMsg = genericErrorMsg + "are not over the same domain";
  throw new owned AssertionError(errorMsg);
```

**Comment by bradcray** [comment]
I think a change like that would work for the associative case, though it's arguably an open issue whether or not it should for the rectangular case.  Specifically, Chapel generally permits operations between rectangular arrays of the same size/shape even if the domains are different / their indices are different.  But wrapping that kind of check in an additional "isAssociative()" style check (if I've guessed at the naming there correctly) might do it.  Then again, there could be other things further downstream that run into troubles for associative domains/arrays if they weren't in the contributor's mind who wrote this package.

**Comment by lucaferranti** [comment]
> it's arguably an open issue whether or not it should for the rectangular case.
Yeah good point, for the rectangular case the domain can be thought as an implementation detail and hence should not affect equality? During testing, sometimes one probably also wants to check that the domain matches the expected one? Sure one can separately test for domain equality if they want to... really not sure what to do :D

for the associative domain case, I think we can agree one should check domain equality, since this would also be consisent with the current definition of `==`

**Comment by lucaferranti** [comment]
regarding the rectangular domain discussion, I think the following is good to notice:

```
use UnitTest;

var c = [1, 2];
var d = [1.0, 2.0];

writeln(c.equals(d));

proc test1(test: borrowed Test) throws {
  test.assertEqual(c, d);
}

UnitTest.main();
```

```
true
test1()
Flavour: FAIL
======================================================================
AssertionError: assert failed -
'1 2'
and
'1.0 2.0'
are not of same type
----------------------------------------------------------------------
```

so `assertEqual` seems to check whether the two objects are indistinguishable and not just equality. If this is the design choice, then probably one should also check the domains match?

**Comment by bradcray** [comment]
I don't really have enough experience with unit testing libraries to have an intuition as to what should happen in this last case.  Do people expect "assertEquals" to be equivalent to `==` (or `+ reduce ==` in the case of arrays?).  If so, then I'd agree that the case above shouldn't be failing.

**Comment by lydia-duncan** [comment]
I just went through our chats when the UnitTest library was being developed.  There were some references to the Python UnitTest library and some expectation that `==` and `assertEquals` would end up with similar results, though the hope was that assertEquals could be more detailed in the case of failures.  However, I don't know that we looked specifically at how Python handled `assertEquals`, so it'd be good to double check those ended up being aligned

**Comment by lucaferranti** [comment]
Sorry for the slow follow-up I meant to answer to this. Python implementation of `assertEqual` is [here](https://github.com/python/cpython/blob/3.11/Lib/unittest/case.py#L835-L873). I think `assertEqual` should be functionally equivalent to checking `==` (or `&& reduce ==` for ararays), but handle error printing in a nicer way. FWIW, in python both `assertEqual([1, 2], [1.0, 2.0])` and `assert [1, 2] == [1.0, 2.0]` succeed.

In general, I would suggest for Chapel `assertEqual(a, b)` to simply check `a == b` and only after it fails eventually checking other things (size, rank, etc.) to print a more informative error message. I think this has a few benefits

1. **Easier to maintain code** the current implementation had a few challenging corner cases (e.g. the associative arrays issue at the beginning, the inconsintency between `assertEqual(1, 1.0)` and `assertEqual([1, 2], [1.0, 2.0])`). Simply checking for `==` woud greatly simplify the logic. Note that one can still e.g. print size, rank etc. for arrays to debug.

2. **More flexibility**: Currently a single function `testAssertSomething(test: borrowed Test)` behaves as a single test even if it has multiple assertions. This way one could still e.g. checking for types by doing
```chapel
proc testString(test: borrowed Test) throws {
  var a = ...,
        b = ....;

  test.assertEqual(a, b);
  test.assertEqual(a.type, b.type);
 }
 ```
 and it would still be a single stricter test.
 
 The second strategy would be to require `assertEqual` to be a stricter check than `==`, but in that case for consistency I think also `test.assertEqual(1, 1.0)` should fail (and one should probably also check for domain equality in arrays).
 
 I would personally lean towards the first option, because more stricter checks can be easily achieved by adding more asssertions to the function and it's well defined. In the second option, I cannot easily opt out if I want lighter checks and also defining "strict equality" in a consistent way could be a little tricky and we would end up with an endless list of "assertion definition" for every type. This is assuming we don't want to check for bit-level equality.
 
 Thoughts on this?

**Comment by lucaferranti** [comment]
> I cannot easily opt out if I want lighter checks

ok, to be fair, this is an exaggeration. I can easily define my own `Test.assertWhateverIWant` method. Nevertheless, I would still lean towards the first option, it follows the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment)  and makes the code easier to maintain.

**Comment by bradcray** [comment]
@lucaferranti : I generally agree with assertEqual() being a richer form of `==`.  One minor difference alluded to above is that `==` on arrays returns an array of bools in Chapel saying whether each element is equal to the corresponding one; so we'd presumably want to reduce that to a single boolean in these assert forms.  Other than that, your proposal sounds fine to me, and consistent with what Lydia found looking through the old notes, if I'm understanding everything correctly.

**Comment by lucaferranti** [comment]
> One minor difference alluded to above is that == on arrays returns an array of bools in Chapel saying whether each element is equal to the corresponding one; so we'd presumably want to reduce that to a single boolean in these assert forms.

Yes, I agree with this. That's actually what I was after in my first message, but I was too lazy when typing and whenever I wrote "equivalent to `==`", I really meant equivalent to `==` or `& reduce ==`. Sorry for the sloppiness 😅

**Comment by lucaferranti** [comment]
regarding associative arrays. AI find the following a little confusing, what is the rationale behind it?

```chapelvar a = ["a" => 1, "b" => 2],
    b = ["c" => 1, "d" => 2],
    c = ["c" => 2, "d" => 1];

writeln(a == b);
writeln(a == c);
```

```
false false
true true
```

**Comment by bradcray** [comment]
> I find the following a little confusing, what is the rationale behind it?

I'm not sure I can defend it, but I can probably explain it:  For rectangular arrays, the array's indices are ignored when doing comparisons, such that `A[1..3] == A[4..6]` could be written, for example, and just look at the array elements, independent of their indices.  We tend to talk in terms of the arrays needing to have similar size and shape.  My guess is that same principle is being applied here, but arguably inappropriately.

That said, I'm slightly surprised by this, as we tend to think of these promoted operations as being a form of zippered iteration, and I thought we only supported zippered iteration over associative arrays when they had identical domains.  Assignment is sometimes a special-case, but I can similarly apply `+` between these arrays, which feels wrong.  If you'd be open to filing a new bug issue against this, that would be great.
